REPL completion for idents starting with number [ci: last-only]

harpocrates opened the pull request:
This relaxes some of the completion result filtering to allow backticked identifiers such as the following to show up in completion results.

     object Test {
       def `1 world` = ""
     }

The pathological builtin names to guard against all start with `<`, even in their encoded forms (backticked `<` would start with a `$`).

Fixes scala/bug#13076

harpocrates committed:
REPL completion for idents starting with number

This relaxes some of the completion result filtering to allow backticked
identifiers such as the following to show up in completion results.

   object Test {
     def `1 world` = ""
   }

The pathological builtin names to guard against all start with `<`,
even in their encoded forms (backticked `<` would start with a `$`).

Fixes scala/bug#13076

harpocrates commented:
Really not sure how to best test this - I thought I was going crazy for a while because the `askTypeCompletion` presentation test case I added _was_ finding the member. The filtering occurs in a place that doesn't seem to have much coverage...

harpocrates commented:
Since I originally hit this through Metals, here it is working after the change:
<img width="581" alt="Screenshot 2025-01-06 at 9 57 00 PM" src="https://github.com/user-attachments/assets/f91291c6-0aab-4c1e-bcd8-f92c5f4f29f9" />

som-snytt commented:
Jenkins discipline.
```
[error] /home/jenkins/workspace/scala-2.13.x-validate-main/src/interactive/scala/tools/nsc/interactive/Global.scala:24:37: Unused import
[error] import scala.reflect.internal.Chars.isIdentifierStart
[error]                                     ^
[error] one error found
```

som-snytt commented:
test
```
--- a/test/junit/scala/tools/nsc/interpreter/CompletionTest.scala
+++ b/test/junit/scala/tools/nsc/interpreter/CompletionTest.scala
@@ -94,6 +94,8 @@ class CompletionTest {
     checkExact(completer, "object O { private def x_y_z = 1; x_y", "}")("x_y_z")
     checkExact(completer, "object x_y_z; import x_y")("x_y_z")

+    checkExact(completer, "object O { def `1 thing` = 1 }; O.")("1 thing")
+
     checkExact(completer, "object x_y_z { def a_b_c }; import x_y_z.a_b")("a_b_c")

     checkExact(completer, "object X { private[this] def definition = 0; def")("definition")
```

possible tweak
```
--- a/src/interactive/scala/tools/nsc/interactive/Global.scala
+++ b/src/interactive/scala/tools/nsc/interactive/Global.scala
@@ -1191,7 +1191,9 @@ class Global(settings: Settings, _reporter: Reporter, projectName: String = "")
       results.filter { (member: Member) =>
         val symbol = member.sym
         def isStable = member.tpe.isStable || member.sym.isStable || member.sym.getterIn(member.sym.owner).isStable
-        def isJunk = !symbol.exists || symbol.name.isEmpty || !isIdentifierStart(member.sym.name.charAt(0)) // e.g. <byname>
+        def isJunk = !symbol.exists || symbol.name.isEmpty ||
+          symbol.name.charAt(0) == '<' ||
+          symbol.name.charAt(0) == '$' && !isIdentifierStart(symbol.encodedName.charAt(0))
         def nameTypeOk: Boolean = {
           forImport || // Completing an import: keep terms and types.
             symbol.name.isTermName == name.isTermName || // Keep names of the same type
```

harpocrates commented:
> test

Thanks for the test pointer! Updated.

> It may already be a ticket that REPL doesn't add backquotes when required.

It doesn't - that's https://github.com/scala/bug/issues/6919. Metals is already compensating for this for Scala 2 and this is fixed outright in Scala 3.

> I noticed <init> satisfies the test for lead bracket, and obviates the test isConstructor.

Updated to remove the `isConstructor` check - `<init>` and `<clinit>` should both have a `<` as their first character in encoded form.

> `symbol.name.charAt(0) == '$' && !isIdentifierStart(symbol.encodedName.charAt(0))`
> ...
> Probably my probe for dollar isn't meaningful. OP meant we don't want to exclude backquoted `<something>`.

What is this aiming to catch? A symbol starting in `$` will encode to something starting in a dollar sign as well (and `isIdentifierStart` will be true for that). I can't see how the first half of that test can be true without the second half being false.

som-snytt commented:
> What is this aiming to catch?

@harpocrates I meant decoded. `member.symbol.name` is decoded so you can check for `<` in `<init>`. You're right that there's no meaningful test beyond that; I intended something like "only check if lead char is encoded", but that isn't useful.

harpocrates commented:
Anything further I should be doing here, or just waiting for someone to have the cycles to review? I'm not in a rush - just want to make sure I'm not accidentally blocking this.

som-snytt commented:
@harpocrates you realize this is scala 2, right? All the cut-throat action is on scala 3.

Seth will want to know what completion does in dotty repl. I think I added `last-only` in case I pushed something?

Thanks for the contrib. LGTM. Someday someone will try to complete something and it will just work.

harpocrates commented:
I know 3 is where it's at, but this bug impacts scala 2 codebases I work with (and ones that are unfortunately very far from scala 3). They have unusual naming conventions where there are a lot of backticked identifiers and package names often start with numbers. 😱 

Completion in dotty works (I showed what it does in the original https://github.com/scala/bug/issues/13076).

som-snytt commented:
Let me add that I love backticked idents, especially for descriptive test method names. I'm not one to judge. I feel confident that this will make it into 2.13.17, but maybe you need it merged to head because you use nightly builds.

harpocrates commented:
> but maybe you need it merged to head because you use nightly builds

No need - codebases in question are still on 2.12.x so I anyways backported the patch to there for our build. I just don't want to commit _that_ into our build until this PR is merged.

lrytz reviewed (approved):
Thanks to both of you!
